src/vmm/src/guest_config/templates/mod.rs

```diff
@@ -0,0 +1,551 @@
+    use std::str::FromStr;
+
+    use serde::de::Error as SerdeError;
+    use serde::{Deserialize, Deserializer};
+
+    use crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags;
+    use crate::guest_config::cpuid::{
+        AmdCpuid, Cpuid, CpuidEntry, CpuidKey, CpuidRegisters, IntelCpuid,
+    };
+    use crate::guest_config::templates::{BitMask, Error, RegisterModifier};
+    use crate::guest_config::x86_64::X86_64CpuConfiguration;
+
+    /// CPUID register enumeration
+    #[allow(unused)]
+    #[allow(missing_docs)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub enum CpuidRegister {
+        Eax,
+        Ebx,
+        Ecx,
+        Edx,
+    }
+
+    /// Composite type that holistically provides
+    /// the location of a specific register being used
+    /// in the context of a CPUID tree.
+    #[allow(unused)]
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct CpuidRegisterModifier {
+        /// Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub leaf: u32,
+        /// Sub-Leaf value.
+        #[serde(deserialize_with = "deserialize_u32_from_str")]
+        pub subleaf: u32,
+        /// KVM feature flags for this leaf-subleaf.
+        #[serde(deserialize_with = "deserialize_kvm_cpuid_flags")]
+        pub flags: crate::guest_config::cpuid::cpuid_ffi::KvmCpuidFlags,
+        /// CPUID register to be modified by the bitmask.
+        #[serde(deserialize_with = "deserialize_cpuid_register")]
+        pub register: CpuidRegister,
+        /// Mask to be applied to register's value at the address
+        /// provided.
+        pub mask: BitMask,
+    }
+
+    /// Wrapper type to containing x86_64 CPU config modifiers.
+    #[derive(Debug, Deserialize, Eq, PartialEq)]
+    pub struct X86_64CpuTemplate {
+        /// Modifiers for CPUID configuration.
+        #[serde(default)]
+        pub cpuid_modifiers: Vec<CpuidRegisterModifier>,
+        /// Modifiers for model specific registers.
+        #[serde(default)]
+        pub msr_modifiers: Vec<RegisterModifier>,
+    }
+
+    impl X86_64CpuTemplate {
+        /// Makes of a CPU template with modifiers
+        /// to be applied on top of the host configuration
+        /// to generate the guest configuration to be used.
+        pub fn apply_template(
+            self,
+            host_config: X86_64CpuConfiguration,
+        ) -> Result<X86_64CpuConfiguration, Error> {
+            let mut guest_msrs_map: HashMap<u64, u64> = HashMap::new();
+            let mut guest_cpuid_map: BTreeMap<CpuidKey, CpuidEntry>;
+
+            // Get the hash map of CPUID data
+            if host_config.cpuid.amd().is_some() {
+                guest_cpuid_map = host_config.cpuid.amd().unwrap().0.clone();
+            } else if host_config.cpuid.intel().is_some() {
+                guest_cpuid_map = host_config.cpuid.intel().unwrap().0.clone();
+            } else {
+                return Err(Error::Cpuid);
+            }
```

Comment:
nit: might be simplified.
```suggestion
            guest_cpuid_map = match host_config.cpuid {
                Cpuid::Intel(cpuid) => cpuid.0.clone(),
                Cpuid::Amd(cpuid) => cpuid.0.clone(),
            };
```

Reply:
@mattschlebusch This is still relevant in your latest version.

Reply:
[Vm contains the supported configuration](https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/vstate/vm.rs#L146) which is where this data is passed from. Short of making another system call to retrieve a new instance of data, or moving the supported configuration out of Vm, the cloning will be necessary as far as I can tell?

Reply:
From I can see in this PR you store the `guest_cpu_config` in `Vm`, not `host_cpu_config`. So why clone `host_cpu_config` if it is instantly thrown away? 
In any case we can leave this as it is and return to this question in final polishing steps.